test/pettingzoo/tic_tac_toe.py

```diff
@@ -1,232 +1,241 @@
-    # policy.set_eps(1)
-    train_collector.collect(n_step=args.batch_size * args.training_num)
-    # log
-    log_path = os.path.join(args.logdir, 'tic_tac_toe', 'dqn')
-    writer = SummaryWriter(log_path)
-    writer.add_text("args", str(args))
-    logger = TensorboardLogger(writer)
-
-    def save_fn(policy):
-        if hasattr(args, 'model_save_path'):
-            model_save_path = args.model_save_path
-        else:
-            model_save_path = os.path.join(
-                args.logdir, 'tic_tac_toe', 'dqn', 'policy.pth'
-            )
-        torch.save(policy.policies[args.agent_id - 1].state_dict(), model_save_path)
-
-    def stop_fn(mean_rewards):
-        return mean_rewards >= args.win_rate
-
-    def train_fn(epoch, env_step):
-        policy.policies[args.agent_id - 1].set_eps(args.eps_train)
-
-    def test_fn(epoch, env_step):
-        policy.policies[args.agent_id - 1].set_eps(args.eps_test)
-
-    def reward_metric(rews):
-        return rews[:, args.agent_id - 1]
-
-    # trainer
-    result = offpolicy_trainer(
-        policy,
-        train_collector,
-        test_collector,
-        args.epoch,
-        args.step_per_epoch,
-        args.step_per_collect,
-        args.test_num,
-        args.batch_size,
-        train_fn=train_fn,
-        test_fn=test_fn,
-        stop_fn=stop_fn,
-        save_fn=save_fn,
-        update_per_step=args.update_per_step,
-        logger=logger,
-        test_in_train=False,
-        reward_metric=reward_metric
-    )
-
-    return result, policy.policies[args.agent_id - 1]
-
-
-def watch(
-    args: argparse.Namespace = get_args(),
-    agent_learn: Optional[BasePolicy] = None,
-    agent_opponent: Optional[BasePolicy] = None,
-) -> None:
-    env = TicTacToeEnv(args.board_size, args.win_size)
-    policy, optim = get_agents(
-        args, agent_learn=agent_learn, agent_opponent=agent_opponent
-    )
-    policy.eval()
-    policy.policies[args.agent_id - 1].set_eps(args.eps_test)
-    collector = Collector(policy, env, exploration_noise=True)
-    result = collector.collect(n_episode=1, render=args.render)
-    rews, lens = result["rews"], result["lens"]
-    print(f"Final reward: {rews[:, args.agent_id - 1].mean()}, length: {lens.mean()}")
+import argparse
+import os
+from copy import deepcopy
+from typing import Optional, Tuple
+
+import gym
+import numpy as np
+import torch
+from pettingzoo.classic import tictactoe_v3
+from torch.utils.tensorboard import SummaryWriter
+
+from tianshou.data import Collector, VectorReplayBuffer
+from tianshou.env import DummyVectorEnv
+from tianshou.env.pettingzoo_env import PettingZooEnv
+from tianshou.policy import (
+    BasePolicy,
+    DQNPolicy,
+    MultiAgentPolicyManager,
+    RandomPolicy,
+)
+from tianshou.trainer import offpolicy_trainer
+from tianshou.utils import TensorboardLogger
+from tianshou.utils.net.common import Net
+
+
+def get_env():
+    return PettingZooEnv(tictactoe_v3.env())

```

Comment:
Is it possible to directly use tictactoc_v3 env?

Reply:
Do you mean we just write like this?
```python
def get_env():
    return tictactoe_v3.env()
```
I am afraid it's not possible since `tictactoe_v3.env()` has different functions and variable naming and standards, the wrapper is necessary to make it compatible with `tianshou`.

Reply:
But instead, pistonball_v4 can be directly used, isn't it?

Reply:
Yes for all the pettingzoo environments we have to first wrap them.
`pistonball_v4 `also use the wrapper in `get_env`.
```python
def get_env(args: argparse.Namespace = get_args()):
    return PettingZooEnv(pistonball_v4.env(continuous=False, n_pistons=args.n_pistons))
```

Reply:
Oh okay, my bad